src/LinuxTracing/UprobesUnwindingVisitor.cpp

```diff
@@ -506,57 +575,94 @@ void UprobesUnwindingVisitor::Visit(uint64_t event_timestamp, const MmapPerfEven
-      std::shared_ptr<unwindstack::MapInfo> map_info = added_map_info->prev_map();
-      while (map_info != nullptr) {
-        if (!map_info->name().empty() && map_info->name().c_str()[0] != '[') {
-          module_path = map_info->name();
-          break;
-        }
-        map_info = map_info->prev_map();
-      }
+  const std::shared_ptr<unwindstack::MapInfo> added_map_info =
+      current_maps_->Find(event_data.address);
+  ORBIT_CHECK(added_map_info != nullptr);  // This is the mapping added with AddAndSort above.
+
+  const std::shared_ptr<unwindstack::Memory> process_memory =
+      unwindstack::Memory::CreateProcessMemory(event_data.pid);
+  auto* const pe = dynamic_cast<unwindstack::PeCoff*>(
+      added_map_info->GetObject(process_memory, unwindstack::ARCH_X86_64));
+
+  // If this is an anonymous executable mapping, we verify whether it belongs to a section of a PE
+  // that was mapped anonymously by Wine because its alignment doesn't obey the requirements of
+  // mmap. If this is not the case, we don't try to send a ModuleUpdateEvent because of this map.
+  if (event_data.filename.empty() && pe == nullptr) {
+    return;
+  }
+
+  std::shared_ptr<unwindstack::MapInfo> closest_file_map_info;
+  if (!event_data.filename.empty()) {
+    closest_file_map_info = added_map_info;
+  } else {
+    // This anonymous executable map corresponds to a PE. We know that this was detected from the
+    // previous file mapping, see MapInfo::GetFileMemoryFromAnonExecMapIfPeCoffTextSection. Find
+    // such mapping and get its file. Note that this assumes that at least the headers of the PE
+    // are already mapped (with a non-executable file mapping).
+    closest_file_map_info = FindFileMapInfoPrecedingAnonMapInfo(added_map_info);
+    if (closest_file_map_info == nullptr) {
+      ORBIT_ERROR("No file mapping found preceding anon exec map at %#x-%#x that belongs to a PE",
+                  added_map_info->start(), added_map_info->end());
+      return;
     }
   }
 
-  if (module_path.empty()) return;
+  // Find the first file mapping with the same name (file path) as the file mapping we just found.
+  // For ELF files, this should correspond to (the first part of) the first loadable segment.
+  // For PEs, this should correspond to the headers.
+  const std::shared_ptr<unwindstack::MapInfo> first_map_info_for_module =
+      FindFirstMapInfoForSameFile(closest_file_map_info);
+  const std::string module_path = closest_file_map_info->name();
+
+  // We want to find the first and the last executable map for this file so that we can create a
+  // ModuleUpdateEvent that encompasses all of them.
+  uint64_t min_exec_map_start = std::numeric_limits<uint64_t>::max();
+  uint64_t max_exec_map_end = 0;
+  {
+    std::shared_ptr<unwindstack::MapInfo> map_info_it = first_map_info_for_module;
+    // Scan the maps forward. Stop when a file mapping for a different file is found.
+    while (map_info_it != nullptr &&
```

Comment:
Should this also be extracted to a method?

Reply:
Fair. Done.